File: ts-core/src/net/neighbor/service.ts
import { Address, Frame, LinkSendError, LinkService, LinkSocket, Protocol } from "@core/net/link";
import { BufferReader, BufferWriter } from "@core/net/buffer";
import { Cost, NodeId, LocalNodeService } from "@core/net/node";
import { NeighborNode, NeighborTable } from "./table";
import { FrameType, GoodbyeFrame, HelloFrame, NeighborFrame } from "./frame";
import { Ok, Result } from "oxide.ts";
import { NotificationService } from "@core/net/notification";
import { CancelListening } from "@core/event";

export class NeighborService {
    #notificationService: NotificationService;
    #localNodeService: LocalNodeService;

    #neighbors: NeighborTable = new NeighborTable();
    #socket: LinkSocket;

    constructor(args: {
        notificationService: NotificationService;
        linkService: LinkService;
        localNodeService: LocalNodeService;
    }) {
        this.#notificationService = args.notificationService;
        this.#localNodeService = args.localNodeService;
        this.#socket = args.linkService.open(Protocol.RoutingNeighbor);
        this.#socket.onReceive((frame) => this.#onFrameReceived(frame));
    }

    onNeighborAdded(listener: (neighbor: Readonly<NeighborNode>) => void): CancelListening {
        return this.#neighbors.onNeighborAdded(listener);
    }

    onNeighborRemoved(listener: (neighborId: NodeId) => void): CancelListening {
        return this.#neighbors.onNeighborRemoved(listener);
    }

    async #onFrameReceived(frame: Frame): Promise<void> {
        const resultNeighborFrame = NeighborFrame.deserialize(frame.reader);
        if (resultNeighborFrame.isErr()) {
            console.warn(`NeighborService: failed to deserialize frame with error: ${resultNeighborFrame.unwrapErr()}`);
            return;
        }

        const neighborFrame = resultNeighborFrame.unwrap();
        if (neighborFrame.type === FrameType.Goodbye) {
            this.#neighbors.removeNeighbor(neighborFrame.senderId);
            this.#notificationService.notify({ type: "NeighborRemoved", nodeId: neighborFrame.senderId });
            return;
        }

        this.#neighbors.addNeighbor(neighborFrame.source.nodeId, neighborFrame.linkCost, frame.remote);
        if (neighborFrame.type === FrameType.Hello) {
            await this.#replyHelloAck(neighborFrame, frame.remote);
        }
        this.#notificationService.notify({
            type: "NeighborUpdated",
            neighborId: neighborFrame.source.nodeId,
            neighborCost: neighborFrame.nodeCost,
            linkCost: neighborFrame.linkCost,
        });
    }

    #sendFrame(frame: HelloFrame | GoodbyeFrame, destination: Address): Result<void, LinkSendError> {
        const writer = new BufferWriter(new Uint8Array(frame.serializedLength()));
        frame.serialize(writer);
        return this.#socket.send(destination, new BufferReader(writer.unwrapBuffer()));
    }

    async sendHello(destination: Address, linkCost: Cost): Promise<Result<void, LinkSendError>> {
        const { source, cost: nodeCost } = await this.#localNodeService.getInfo();
        return this.#sendFrame(new HelloFrame({ type: FrameType.Hello, source, nodeCost, linkCost }), destination);
    }

    async #replyHelloAck(frame: HelloFrame, destination: Address): Promise<Result<void, LinkSendError>> {
        const { source, cost: nodeCost } = await this.#localNodeService.getInfo();
        return this.#sendFrame(
            new HelloFrame({ type: FrameType.HelloAck, source, nodeCost, linkCost: frame.linkCost }),
            destination,
        );
    }

    async sendGoodbye(destination: NodeId): Promise<Result<void, LinkSendError>> {
        const frame = new GoodbyeFrame({ senderId: await this.#localNodeService.getId() });
        const addrs = this.#neighbors.getNeighbor(destination)?.addresses;
        this.#neighbors.removeNeighbor(destination);
        if (addrs !== undefined && addrs.length > 0) {
            return this.#sendFrame(frame, addrs[0]);
        } else {
            return Ok(undefined);
        }
    }

    async resolveAddress(id: NodeId): Promise<Address[]> {
        const convertedId = await this.#localNodeService.convertLocalNodeIdToLoopback(id);
        return this.#neighbors.getAddresses(convertedId);
    }

    hasNeighbor(id: NodeId): boolean {
        return this.#neighbors.getNeighbor(id) !== undefined;
    }

    getNeighbor(id: NodeId): NeighborNode | undefined {
        return this.#neighbors.getNeighbor(id);
    }

    getNeighbors(): NeighborNode[] {
        return this.#neighbors.getNeighbors();
    }
}

File: py-core/py_core/net/neighbor/service.py
import asyncio
from dataclasses import dataclass
from enum import Flag

from py_core.net.link.address import Address
from py_core.net.link.frame import Protocol, ProtocolValue
from py_core.net.link.service import LinkService, LinkSocket
from py_core.net.local.service import LocalNodeService
from py_core.net.neighbor.table import (
    NeighborNode,
    NeighborTable,
    NeighborTableListener,
)
from py_core.net.node import NodeId, Cost
from py_core.net.notification.service import NeighborRemoved, NeighborUpdated
from py_core.serde.buffer import BufferWriter
from py_core.serde.derive import DeriveDeserialize, DeriveSerialize
from py_core.serde.enum import EnumSerde
from py_core.net.notification import NotificationService


class NeighborControlFlagValue(Flag):
    Empty = 0
    KeepAlive = 1


@dataclass(frozen=True)
class NeighborControlFlag(EnumSerde[NeighborControlFlagValue]):
    @staticmethod
    def enum_type():
        return NeighborControlFlagValue


@dataclass(frozen=True)
class NeighborControlFrame(DeriveDeserialize, DeriveSerialize):
    flags: NeighborControlFlag
    source_node_id: NodeId
    link_cost: Cost

    def should_reply_immediately(self) -> bool:
        return (self.flags.value & NeighborControlFlagValue.KeepAlive) == 0


class NeighborService:
    _notification_service: NotificationService
    _local_node_service: LocalNodeService

    _neighbor_table: NeighborTable
    _socket: LinkSocket

    def __init__(
        self,
        link_service: LinkService,
        notification_service: NotificationService,
        local_node_service: LocalNodeService,
    ):
        self._notification_service = notification_service
        self._local_node_service = local_node_service

        self._neighbor_table = NeighborTable()
        self._socket = link_service.open(Protocol(ProtocolValue.RoutingNeighbor))

        this = self

        class Listener(NeighborTableListener):
            def on_neighbor_updated(self, neighbor: NeighborNode) -> None:
                notification = NeighborUpdated(neighbor.node_id, neighbor.link_cost)
                notification_service.notify(notification)

            def on_neighbor_removed(self, neighbor: NeighborNode) -> None:
                notification = NeighborRemoved(neighbor.node_id)
                notification_service.notify(notification)

            def on_neighbor_should_send_hello(self, neighbor: NeighborNode) -> None:
                address = next(iter(neighbor.addresses))
                flags = NeighborControlFlag(NeighborControlFlagValue.Empty)
                asyncio.create_task(
                    this._send_hello(address, neighbor.link_cost, flags)
                )

        self._neighbor_table.add_listener(Listener())

    async def _send_hello(
        self, address: Address, link_cost: Cost, flags: NeighborControlFlag
    ) -> None:
        frame = NeighborControlFrame(
            flags=flags,
            source_node_id=await self._local_node_service.node_id(),
            link_cost=link_cost,
        )

        writer = BufferWriter(frame.serialized_length())
        frame.serialize(writer)
        await self._socket.send(address, writer.unwrap_buffer())

    async def send_hello(self, address: Address, link_cost: Cost) -> None:
        await self._send_hello(
            address, link_cost, NeighborControlFlag(NeighborControlFlagValue.Empty)
        )

    def get_neighbor(self, node_id: NodeId) -> NeighborNode | None:
        return self._neighbor_table.get_neighbor(node_id)

    def resolve_neighbor_from_address(self, address: Address) -> NeighborNode | None:
        return self._neighbor_table.resolve_neighbor_from_address(address)

    def on_frame_sent(self, destination: NodeId) -> None:
        self._neighbor_table.delay_send_hello(destination)

    def on_frame_received(self, source: NodeId) -> None:
        self._neighbor_table.delay_expire(source)
